src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1020,15 +1022,17 @@ private class RsFnInferenceContext(
             } else {
                 val parameters = callee.element.typeParameterList?.typeParameterList.orEmpty()
                     .map { TyTypeParameter.named(it) }
-                parameters.zip(typeArguments).toMap()
+                parameters.zip(typeArguments).toMap().toTypeSubst()
             }
         }
 
         unifySubst(fnSubst, typeParameters)
 
         val methodType = (callee.element.typeOfValue)
             .substitute(typeParameters)
-            .foldWith(this::normalizeAssociatedTypesIn) as TyFunction
+            .foldWith(object: TypeFolder {
+                override fun foldTy(type: Ty): Ty = normalizeAssociatedTypesIn(type)
```

Comment:
This code occurs in multiple places. Maybe it's better to extract `AssociatedTypeNormalizer` to a class level?